src/manage_nsfs/nc_lifecycle.js

```diff
@@ -908,7 +870,437 @@ class NCLifecycle {
                 }
             }
         }
-}
+    }
+
+    /**
+     * _set_rule_state sets the current rule state on the lifecycle run status
+     * @param {Object} bucket_json 
+     * @param {*} lifecycle_rule 
+     * @param {{is_finished?: Boolean | Undefined, candidates_file_offset?: number | undefined}} rule_state 
+     * @returns {Void}
+     */
+    _set_rule_state(bucket_json, lifecycle_rule, rule_state) {
+        this.lifecycle_run_status.buckets_statuses[bucket_json.name].rules_statuses[lifecycle_rule.id].state = rule_state;
+    }
+
+    /**
+     * _get_rule_sate gets the current rule state on the lifecycle run status
+     * @param {Object} bucket_json 
+     * @param {*} lifecycle_rule 
+     * @returns {{is_finished?: Boolean | Undefined, candidates_file_offset?: number | undefined}} rule_state 
+     */
+    _get_rule_sate(bucket_json, lifecycle_rule) {
+        return this.lifecycle_run_status.buckets_statuses[bucket_json.name].rules_statuses[lifecycle_rule.id]?.state;
```

Comment:
I don't think this function should return undefined. first state should already be initialized at this point. if its possible it isn't undefined it should return a new object